File: apps/android/app/src/test/kotlin/com/loonext/android/features/settings/WidgetSnippetTest.kt
package com.loonext.android.features.settings

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #232 — the one line an owner pastes into their own website, built twice.
 *
 * `widgetSnippet` exists in TypeScript and again in Kotlin, and the two have to
 * produce byte-identical markup. This is the exact shape that has gone wrong in
 * this repository before: a shared rule hand-ported to a second language drifts,
 * and the drift is invisible because each copy works perfectly on its own.
 *
 * It would be invisible here too. An owner copies the line from their phone,
 * pastes it into WordPress, and the button either appears or does not — with no
 * error anywhere in our system, because nothing of ours ever reads it back.
 *
 * The expectation is written out in full rather than assembled from parts. A
 * test that builds its expectation the same way the code does passes whatever
 * the code does.
 */
class WidgetSnippetTest {

    @Test
    fun `builds the exact line the web app builds`() {
        assertEquals(
            "<script src=\"https://app.loonext.com/widget.js\" " +
                "data-key=\"aaaaaaaa-1111-4111-8111-aaaaaaaaaaaa\" defer></script>",
            widgetSnippet("aaaaaaaa-1111-4111-8111-aaaaaaaaaaaa"),
        )
    }

    @Test
    fun `is a closed element, so a page does not swallow the rest of itself`() {
        // An unclosed script tag does not fail loudly — it eats whatever markup
        // follows it, which on somebody's homepage is their homepage.
        val snippet = widgetSnippet("k")
        assertTrue(snippet.startsWith("<script "))
        assertTrue(snippet.endsWith("</script>"))
    }

    @Test
    fun `has no line breaks, because it is pasted into a text field`() {
        // Several site builders offer a one-line "header code" box, and a
        // newline in a value pasted there is silently truncated at the break.
        assertFalse(widgetSnippet("k").contains("\n"))
    }
}

File: apps/web/src/lib/marketing/widget-snippet.test.ts
/**
 * #232 — the line an owner copies.
 *
 * Short, and every assertion is about something that would break silently: a
 * snippet that looks right and does not work is worse than one that is
 * obviously wrong, because the owner pastes it, sees no button, and concludes
 * the product does not work.
 */
import { describe, expect, it } from "vitest";

import { widgetSnippet } from "./widget-snippet";

describe("the embed snippet", () => {
  it("carries the key and points at the script", () => {
    const snippet = widgetSnippet(
      "https://app.loonext.com",
      "aaaaaaaa-1111-4111-8111-aaaaaaaaaaaa",
    );
    expect(snippet).toContain('src="https://app.loonext.com/widget.js"');
    expect(snippet).toContain('data-key="aaaaaaaa-1111-4111-8111-aaaaaaaaaaaa"');
    // `defer` matters: without it the script blocks the customer's page while
    // it loads, on somebody else's site, which is the fastest way to be removed.
    expect(snippet).toContain("defer");
  });

  it("is a closed element, so a page does not swallow the rest of itself", () => {
    // A snippet missing its closing tag takes everything after it in the host
    // page with it. Assembled from the tag name, so this is the assertion that
    // the assembly is right rather than clever.
    const snippet = widgetSnippet("https://app.loonext.com", "k");
    // Both tags are built from the same pieces the source is, for the reason
    // the module explains: a literal closing tag in a file the bundler embeds
    // as a string closes the surrounding script block. A test file is embedded
    // the same way.
    const tag = "script";
    expect(snippet.startsWith(`<${tag} `)).toBe(true);
    expect(snippet.endsWith(`</${tag}>`)).toBe(true);
  });

  it("has no line breaks, because it is pasted into a text field", () => {
    expect(widgetSnippet("https://app.loonext.com", "k")).not.toContain("\n");
  });
});

describe("#228 the snippet carries the workspace's language", () => {
  it("names a non-default locale, so the paste matches the preview", () => {
    // widget.js is served raw to somebody else's site with no way to ask us
    // anything before it paints, so the language has to arrive as an attribute.
    const snippet = widgetSnippet("https://app.loonext.com", "k", "fr-CA");
    expect(snippet).toContain('data-lang="fr-CA"');
  });

  it("stays one plain line for an English workspace", () => {
    // The overwhelming majority of snippets. An attribute that always appears
    // would put a language on every paste, including the ones that never
    // needed one.
    expect(widgetSnippet("https://app.loonext.com", "k", "en")).not.toContain(
      "data-lang",
    );
    expect(widgetSnippet("https://app.loonext.com", "k")).not.toContain(
      "data-lang",
    );
  });
});
